### PR TITLE
Devel

### PR DESCRIPTION
**Pull request by senier**

### CODE REVIEW

File: capdpa/IR/ir_type.py [hunk 1]
```diff
@@ -12,6 +12,12 @@ def __init__(self, name, pointer = 0, constant = False, reference = False, **kwa
         self.reference = reference
         self.constant = constant
 
+    def UsedTypes(self, parent):
+        fqn = self.FullyQualifiedName()
+        if parent == fqn[:-1] or (parent == fqn[:-2] and fqn[-1] == "Class"):
+            return []
+        return [self.FullyQualifiedName()]
```
Comment (jklmnn) [inline capdpa/IR/ir_type.py:19]: I'd put the second return in an else block:
```Python
    if parent == fqn[:-1] or (parent == fqn[:-2] and fqn[-1] == "Class"):
        return []
    else
        return [self.FullyQualifiedName()]
```
I think this is more clear (especially if the function grows). There was a talk about this matter, this was the preferred solution and I'd go with the argumentation they made (I can't rephrase it but I'll link the talk when I find it again).
I'm aware that we probably don't follow this style in our code base but I'd advocate to do so.

File: capdpa/IR/ir_unit.py [hunk 2]
```diff
@@ -21,6 +25,11 @@ def IsPrivate(self):
     def InstantiateTemplates(self):
         pass
 
+    def UsedTypes(self, parent):
+        if parent == self.name.PackageFull()[:-1]:
+            return []
+        return [self.name.PackageFull()]
```
Comment (jklmnn) [inline capdpa/IR/ir_unit.py:31]: same as above

File: capdpa/IR/ir_unit.py [hunk 3]
```diff
@@ -32,20 +41,5 @@ def PackagePath(self):
...
-                types.append(f.return_type)
-
-        map(lambda v: types.append(v.ctype), filter(ir_variable.Member.isInst, self.children))
-        map(lambda t: types.append(t.reference), filter(lambda c: ir_type.Type_Definition.isInst(c) and c.reference, self.children))
-        map(lambda c: types.append(c), filter(Class_Reference.isInst, self.children))
-        map(lambda t: types.append(t.ctype), filter(ir_variable.Variable.isInst, self.children))
-
-        return sorted(list(set(map(lambda t: t.PackageName(), filter(isLocalType, types)))))
+        packages = map(lambda x: x, self.UsedTypes(self.FullyQualifiedName()))
+        return sorted(list(set([".".join (map(self.ConvertName, p)) for p in packages])))
```
Comment (jklmnn) [inline capdpa/IR/ir_unit.py]: Ada space after `join`? I'd remove it in Python.

File: capdpa/IR/ir_unit.py [hunk 4]
```diff
@@ -32,20 +41,5 @@ def PackagePath(self):
...
-            if f.return_type:
-                types.append(f.return_type)
-
-        map(lambda v: types.append(v.ctype), filter(ir_variable.Member.isInst, self.children))
-        map(lambda t: types.append(t.reference), filter(lambda c: ir_type.Type_Definition.isInst(c) and c.reference, self.children))
-        map(lambda c: types.append(c), filter(Class_Reference.isInst, self.children))
-        map(lambda t: types.append(t.ctype), filter(ir_variable.Variable.isInst, self.children))
-
-        return sorted(list(set(map(lambda t: t.PackageName(), filter(isLocalType, types)))))
+        packages = map(lambda x: x, self.UsedTypes(self.FullyQualifiedName()))
```
Comment (jklmnn) [inline capdpa/IR/ir_unit.py]: Isn't `map(lambda x: x, <list>)` equivalent to just `<list>`?

File: tests/capdpa_test.py [hunk 5]
```diff
@@ -10,11 +10,13 @@ def check(self, a, b):
...
 
-    def cache(self, res, spec):
-        if not os.path.exists("tests/cache/"):
-            os.mkdir("tests/cache/")
-        return open("tests/cache/" + spec, "w").write(res)
+    def cache(self, res, spec, directory=""):
+        if not os.path.exists("tests/cache"):
+            os.mkdir("tests/cache")
+        if not os.path.exists("tests/cache/" + directory):
+            os.mkdir("tests/cache/" + directory)
```
Comment (jklmnn) [inline tests/capdpa_test.py]: The check for `tests/cache` could be removed if we use `os.makedirs` instead as it works recursively.

File: tests/testgenerator.py [hunk 6]
```diff
@@ -325,19 +295,16 @@ def test_pointer_depth(self):
...
     def test_spec_file_name(self):
         result = Namespace(name = "Capdpa", children = [
             Class(name = "Child1"),
             Namespace(name = "Child2", children = [
                 Class(name = "Grandchild")])]).AdaSpecification()
         self.check(result[0].FileName(), "capdpa.ads")
-        self.check(result[1].FileName(), "capdpa-child1.ads")
-        self.check(result[2].FileName(), "capdpa-child2.ads")
-        self.check(result[3].FileName(), "capdpa-child2-grandchild.ads")
+        self.check(result[1].FileName(), "capdpa-child2.ads")
```
Comment (jklmnn) [inline tests/testgenerator.py:308]: Maybe we should extend this test again to create a grand child package again? The initial goal of this test was to test file name generation for different depths of child packages.

**Review by jklmnn — CHANGES_REQUESTED** [review]
Generally we should think about renaming the `.txt` Ada test files to `.ads` as they're specs and the current setting is inconsistent.

**Comment by senier** [comment]
IMHO we should address the style and structural issues more generally. What would be be a good tool to automatically enforce and check a certain style (including spaces before brackets, if-else, ...)?

**Comment by jklmnn** [comment]
I have only used [pylint](https://www.pylint.org/) which did a good job. It's also quite configurable so we can adapt it to our needs.

File: capdpa/IR/ir_unit.py (hunk 4 above)
Comment (senier) [reply capdpa/IR/ir_unit.py]: Good catch! There was an operation on `x` in an earlier version. Removed.

File: capdpa/IR/ir_unit.py (hunk 3 above)
Comment (senier) [reply capdpa/IR/ir_unit.py]: Done.

File: tests/capdpa_test.py (hunk 5 above)
Comment (senier) [reply tests/capdpa_test.py]: That's much nicer, indeed. Changed.

File: tests/testgenerator.py (hunk 6 above)
Comment (senier) [reply tests/testgenerator.py:308]: Good point. Done.

File: capdpa/IR/ir_type.py (hunk 1 above)
Comment (senier) [reply capdpa/IR/ir_type.py:19]: TBH, you're argument is not very convincing ;-) Are you okay with deferring this discussion until we have a coding style and style-checking tools in place?

**Comment by senier** [comment]
@treiher [uses pylint and flake8](https://github.com/Componolit/RecordFlux/blob/master/Makefile) in RecordFlux. We should adopt this for Cappulada (cf. #68).

**Review by jklmnn — APPROVED** [review]

**Pull request merged**